include/lo2s/perf/event_provider.hpp

```diff
@@ -0,0 +1,96 @@
+/*
+ * This file is part of the lo2s software.
+ * Linux OTF2 sampling
+ *
+ * Copyright (c) 2017,
+ *    Technische Universitaet Dresden, Germany
+ *
+ * lo2s is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * lo2s is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with lo2s.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include <stdexcept>
+#include <string>
+#include <unordered_map>
+
+#include "lo2s/platform.hpp"
+
+namespace lo2s
+{
+namespace perf
+{
+
+class EventProvider
+{
+public:
+    using EventMap = std::unordered_map<std::string, platform::CounterDescription>;
+
+    EventProvider();
+    EventProvider(const EventProvider&) = delete;
+    void operator=(const EventProvider&) = delete;
+
+    static const inline EventProvider& instance()
```

Comment:
Please note, according to the C++ standard, class member functions _defined_ inside of a class definition are implicitly inline. (§ 9.3.2 [class.mfct])